### ISSUE TITLE
Bug with interpolation near edges

### ISSUE THREAD

**Issue by adler-j** [odlgroup/odl#709]
With `interp='nearest'`, the result is very reasonable:

```python
>>> space = odl.uniform_discr(0, 1, 5, interp='nearest')
>>> x = space.one()
>>> x.interpolation(1)
1.0
```

But with `interp='linear'`, the result seems very off:

```python
>>> space = odl.uniform_discr(0, 1, 5, interp='linear')
>>> x = space.one()
>>> x.interpolation(1)
0.5
```

Thanks @chongchenmath for pointing this out!

**Comment by adler-j** [comment]
I tried to quick-fix this, but there is a few too many layers of code for me to digest quickly, perhaps @kohr-h could do it. It would also be nice if the utility functions used when doing the interpolation like `_create_weight_edge_lists` and `_compute_linear_weights_edge` could use some documentation since I have a really hard time following them without going in and adding print statements to see what is up.

**Comment by chongchenmath** [comment]
If the intensity values of image on the boundary are 0, then there is no problem when we deform the image.  Otherwise, the boundary problem might be appeared. @adler-j @kohr-h For example,when we update a deformation by a linear deformation, each component has a non-zero boundary.

**Comment by kohr-h** [comment]
Yes, there is clearly an issue. The value in the linear interpolation case should be 0.

**Comment by adler-j** [comment]
Why should it be zero? Wouldn't next to all users expect 1 in this case?

As an argument for this, consider the case of

```
>>> space = uniform_discr(0, 1, 1, interp='linear')
```

Here we have, as expected by basically any user:

```
>>> space.one().norm()
1.0
```

But if we use the definition as you have given it, (0 at border) we would expect

```
space.one().norm() = \int_0^0.5 (2x)^2 dx + \int_0.5^1 (1-2x)^2 dx = 1/3
```

Right? But perhaps it's the norm that is wrong here?

Anyway this is all due to #179 lacking, we could formalize it through that.

**Comment by kohr-h** [comment]
Right, formalizing through solving #179 would be the proper way to do it. But it also has a mathematical component, as you pointed out. We have, in principle, 3 competing goals:

- Basis functions add up to 1 everywhere
- Interpolator is continuous everywhere
- Boundary values are attained in a "natural" way.

I'll put some more thoughts into this.

**Comment by kohr-h** [comment]
Having written this, I should also mention the current status. It's perhaps a bit ad-hoc but basically the last function is centered at the last grid point, which is 0.9 in the above example (similar for left boundary). So if you put a hat there, It has no chance of giving value 1 at the boundary. For that, you would need a "half" hat that extends to the boundary and then sharply drops to 0. Which makes your function discontinuous.

**Comment by adler-j** [comment]
> Which makes your function discontinuous.

It's only discontinuous if you assume it extends outside, and even so you could imagine something like this:

![basisfuncs](https://cloud.githubusercontent.com/assets/2202312/20130253/d8b655d2-a655-11e6-8d93-93a70a23fd90.png)

where the basis functions are color coded.

**Comment by kohr-h** [comment]
Nice picture :-)
Except that the basis functions don't extend beyond the neighboring grid point.

Anyway, I'll make an attempt at solving this issue.

**Comment by adler-j** [comment]
> Except that the basis functions don't extend beyond the neighboring grid point.

Well then you need to add that to your list of constraints ;)

**Comment by adler-j** [comment]
Has any work been done on this @kohr-h ?

**Comment by kohr-h** [comment]
Yes, I actually have worked out the details and started implementing things back in November, see [this branch](https://github.com/kohr-h/odl/tree/issue-709__interpolation_bdry). Some hairy details are yet to be sorted out though.

**Comment by adler-j** [comment]
It is quite critical to get this done for our work on deformations. Some functions (specifically the coordinate function) are not of compact support and currently breaks badly.

**Comment by adler-j** [comment]
Any chance you could get this fixed @kohr-h? Even a partial fix would go a long way since this is a hairy bug that puts people off.

**Comment by kohr-h** [comment]
I'll see what I can do. It's not quite trivial to get this to run fast, though, and it's tedious detail work. My guess is that this would require me one full day with nothing else. I'll put it on my short list.
